narwhals/_pandas_like/utils.py

```diff
@@ -553,52 +558,47 @@ def int_dtype_mapper(dtype: Any) -> str:
     return "int64"
 
 
-def calculate_timestamp_datetime(  # noqa: C901, PLR0912
-    s: NativeSeriesT, original_time_unit: str, time_unit: str
+_TIMESTAMP_DATETIME_OP_FACTOR: Mapping[
+    tuple[UnitCurrent, UnitTarget], tuple[BinOpBroadcast, IntoRhs]
```

Comment:
I think the issue is that `BinOpBroadcast` is generic, but using a `TypeVar` with a default.

So you don't get a warning when using `BinOpBroadcast` *without* specialising - but it causes an issue later when the default conflicts with the constraints of the function signature.

Hopefully this works, I'm replying from my phone 😅
```suggestion
    tuple[UnitCurrent, UnitTarget], tuple[BinOpBroadcast[NativeSeriesT], IntoRhs]
```

Reply:
```py
narwhals/_pandas_like/utils.py:583: error: Argument 1 has incompatible type "_CuDFSeries"; expected "Series[Any]"  [arg-type]
narwhals/_pandas_like/utils.py:583: error: Argument 1 has incompatible type "_ModinSeries"; expected "Series[Any]"  [arg-type]
```

https://github.com/narwhals-dev/narwhals/blob/bb037039f2b97419a1facefa860d2cce6965a60e/narwhals/_pandas_like/typing.py#L25-L31

`NativeSeriesT` means *exactly one of* `pd.Series[Any]`, `_CuDFSeries`, `_ModinSeries`, but default to `pd.Series[Any]` when not provided.

So the conflict is that 
- `NativeSeriesT` is still used for the function parameter and return type
- But `_TIMESTAMP_DATETIME_OP_FACTOR` is currently saying it only accepts `pd.Series[Any]`

https://github.com/narwhals-dev/narwhals/blob/0e6def9aee19e8d693c958184533617acab353f8/narwhals/_pandas_like/utils.py#L576-L578

Reply:
> Hey @dangotbanned thanks for investigating this.
> 
> > But _TIMESTAMP_DATETIME_OP_FACTOR is currently saying it only accepts pd.Series[Any]
> 
> Why would that be the case? The definition is:
> 
> `BinOpBroadcast: TypeAlias = Callable[[NativeSeriesT, int], NativeSeriesT]`
> 
> Regarding [#2838 (comment)](https://github.com/narwhals-dev/narwhals/pull/2838#discussion_r2211640442), `BinOpBroadcast` is not a generic, and it would result in:


`BinOpBroadcast` is a generic [`TypeAlias`](https://typing.python.org/en/latest/spec/aliases.html) **because** it uses a `TypeVar` (`NativeSeriesT`) in it's definition.

<img width="600" alt="image" src="https://github.com/user-attachments/assets/eda7cc17-2182-4c4c-8a1c-7cef5519f250" />


If we compare that to when you hover over the current use, it uses the default type - because it wasn't subscribed 

<img width="600" alt="image" src="https://github.com/user-attachments/assets/a8f813d0-1afb-4c53-a743-d36308dd45fa" />


--- 

> 
> > Type variable "NativeSeriesT" has no meaning in this contextPylance[reportGeneralTypeIssues]
> 
> 🧐


Ahh, so that message it about [scoping](https://typing.python.org/en/latest/spec/generics.html#scoping-rules-for-type-variables) - not that the `TypeVar` can't be used at all


So this is because `_TIMESTAMP_DATETIME_OP_FACTOR` is in a global scope, but we're trying to solve the `TypeVar` in a function scope

Reply:
I've just used `Any` instead in (https://github.com/narwhals-dev/narwhals/pull/2838/commits/bdb46aa1c49eeb933a78131f6a3d2b7032c98c1f)

The main reason I had that typing in `_arrow` was because the stubs caused too many false positives when I used the functions directly.

Here, we're just using `operator` functions, which are mostly typed with `Any` *anyway* 😉 

https://github.com/python/typeshed/blob/8e25bda401332ac56d4750cffda2127a4302cd2f/stdlib/_operator.pyi#L42-L71

Reply:
We still have `NativeSeriesT` preserved on the outside, so this is simpler than refactoring everything to provide a scope for the `TypeVar` 😅